atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionManager.java

```diff
@@ -64,25 +64,21 @@ protected final <T, E extends Exception> T runTaskThrowOnConflict(TransactionTas
     protected final <T, E extends Exception> T runTaskThrowOnConflictWithCallback(
             TransactionTask<T, E> task, CallbackAwareTransaction txn, Runnable callback)
             throws E, TransactionFailedException {
+        checkOpen();
         try {
-            checkOpen();
-            try {
-                T ret = task.execute(txn);
-                if (txn.isUncommitted()) {
-                    txn.commitWithoutCallbacks();
-                }
-                return ret;
-            } finally {
-                // Make sure that anyone trying to retain a reference to this transaction
-                // will not be able to use it.
-                if (txn.isUncommitted()) {
-                    txn.abort();
-                }
+            txn.onCommitOrAbort(txn::reportExpectationsCollectedData);
```

Comment:
Is this not a break in behavior? Previously this was in a finally here, now you're only going to run it if someone calls #abort or #commit?

Reply:
Nevermind, it's the same. I'm getting really confused with callbacks and onCommitOrAbort

Reply:
Yes: specifically, if task.execute throws, we'll call abort in the finally here, otherwise we'll proceed to try to commit. SO this is all nice and safe.